File: app/src/main/java/com/dammaj/app/data/WebScraper.java
package com.dammaj.app.data;

import android.util.Log;

import com.dammaj.app.model.ContentItem;

import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class WebScraper {
    private static final String TAG = "WebScraper";
    private final ExecutorService executorService;

    public WebScraper() {
        this.executorService = Executors.newFixedThreadPool(2);
    }

    public void scrapeHome(ScraperCallback callback) {
        executorService.execute(() -> {
            try {
                List<ContentItem> items = fetchFromHome();
                callback.onSuccess(items);
            } catch (Exception e) {
                Log.e(TAG, "Error scraping home", e);
                callback.onFailure(e);
            }
        });
    }

    public void scrapeCategory(String category, ScraperCallback callback) {
        executorService.execute(() -> {
            try {
                List<ContentItem> items = fetchFromCategory(category);
                callback.onSuccess(items);
            } catch (Exception e) {
                Log.e(TAG, "Error scraping category: " + category, e);
                callback.onFailure(e);
            }
        });
    }

    private List<ContentItem> fetchFromHome() throws IOException {
        List<ContentItem> items = new ArrayList<>();

        try {
            String url = "https://mimotoufik11-stack.github.io/Toufiko/";
            Document doc = connectWithUserAgent(url).get();

            items = parseDocument(doc);

            if (items.isEmpty()) {
                Log.w(TAG, "No items found using primary selector, trying fallback");
                items = parseFallback(doc);
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to fetch from home", e);
            throw e;
        }

        return items;
    }

    private List<ContentItem> fetchFromCategory(String category) throws IOException {
        List<ContentItem> items = new ArrayList<>();

        try {
            String url = "https://mimotoufik11-stack.github.io/Toufiko/?category=" + category;
            Document doc = connectWithUserAgent(url).get();

            items = parseDocument(doc);

            if (items.isEmpty()) {
                items = parseFallback(doc);
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to fetch category: " + category, e);
            throw e;
        }

        return items;
    }

    private List<ContentItem> parseDocument(Document doc) {
        List<ContentItem> items = new ArrayList<>();

        try {
            Elements articles = doc.select("article, .content-item, [data-type='content']");

            if (articles.isEmpty()) {
                Log.w(TAG, "No articles found with main selectors");
                return items;
            }

            for (Element article : articles) {
                try {
                    ContentItem item = extractContentItem(article);
                    if (item != null && item.getTitle() != null && !item.getTitle().isEmpty()) {
                        items.add(item);
                    }
                } catch (Exception e) {
                    Log.w(TAG, "Error parsing individual item", e);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Error parsing document", e);
        }

        return items;
    }

    private List<ContentItem> parseFallback(Document doc) {
        List<ContentItem> items = new ArrayList<>();

        try {
            Elements articles = doc.select("div, section, li");

            for (Element element : articles) {
                try {
                    if (element.hasAttr("data-content") || element.hasAttr("data-item")) {
                        ContentItem item = extractContentItem(element);
                        if (item != null && item.getTitle() != null) {
                            items.add(item);
                        }
                    }
                } catch (Exception e) {
                    Log.w(TAG, "Error parsing fallback item", e);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Error with fallback parsing", e);
        }

        return items;
    }

    private ContentItem extractContentItem(Element element) {
        try {
            ContentItem item = new ContentItem();

            item.setId(element.attr("data-id").isEmpty() ? System.currentTimeMillis() + "" : element.attr("data-id"));

            Element titleElement = element.selectFirst("h1, h2, h3, .title, [data-title]");
            if (titleElement != null) {
                item.setTitle(titleElement.text());
            }

            Element linkElement = element.selectFirst("a");
            if (linkElement != null) {
                item.setLinkUrl(linkElement.attr("href"));
            }

            Element imageElement = element.selectFirst("img");
            if (imageElement != null) {
                item.setImageUrl(imageElement.attr("src"));
            }

            Element typeElement = element.selectFirst("[data-type], .type, .category");
            if (typeElement != null) {
                item.setType(typeElement.text());
            } else {
                item.setType("Content");
            }

            Element descElement = element.selectFirst("p, .description, [data-description]");
            if (descElement != null) {
                item.setDescription(descElement.text());
            }

            Element authorElement = element.selectFirst(".author, [data-author]");
            if (authorElement != null) {
                item.setAuthor(authorElement.text());
            }

            item.setDateAdded(String.valueOf(System.currentTimeMillis()));

            return item;
        } catch (Exception e) {
            Log.e(TAG, "Error extracting content item", e);
            return null;
        }
    }

    private Connection connectWithUserAgent(String url) {
        return Jsoup.connect(url)
                .userAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
                .timeout(10000)
                .followRedirects(true);
    }

    public void shutdown() {
        executorService.shutdown();
    }
}

File: app/src/main/java/com/alilmia/app/data/WebScraper.kt
package com.alilmia.app.data

import com.alilmia.app.model.ContentItem
import com.alilmia.app.model.ContentType
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import java.net.URLEncoder
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Web scraper for Al-Ilmia website using Jsoup
 * Handles background threading and robust error handling
 */
class WebScraper {
    
    private val executor: ExecutorService = Executors.newFixedThreadPool(3)
    private val baseUrl = "https://alilmia.com"
    
    /**
     * Scrape homepage content
     */
    fun scrapeHome(callback: ScraperCallback) {
        executor.execute {
            try {
                val doc = Jsoup.connect(baseUrl)
                    .timeout(30000)
                    .userAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
                    .get()
                
                val items = parseDocument(doc)
                callback.onSuccess(items)
            } catch (e: Exception) {
                callback.onFailure(e)
            }
        }
    }
    
    /**
     * Scrape specific category content
     */
    fun scrapeCategory(category: String, callback: ScraperCallback) {
        executor.execute {
            try {
                val encodedCategory = URLEncoder.encode(category, "UTF-8")
                val url = "$baseUrl/category/$encodedCategory"
                
                val doc = Jsoup.connect(url)
                    .timeout(30000)
                    .userAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
                    .get()
                
                val items = parseDocument(doc, category)
                callback.onSuccess(items)
            } catch (e: Exception) {
                callback.onFailure(e)
            }
        }
    }
    
    /**
     * Parse document to extract content items
     */
    private fun parseDocument(doc: Document, category: String? = null): List<ContentItem> {
        val items = mutableListOf<ContentItem>()
        
        try {
            // Try multiple selectors for robustness
            val articles = doc.select("article, .post, .entry, .content-item")
            
            for (element in articles) {
                try {
                    val item = extractContentFromElement(element, category)
                    if (item != null) {
                        items.add(item)
                    }
                } catch (e: Exception) {
                    // Continue processing other elements
                }
            }
            
            // If no items found, try alternative selectors
            if (items.isEmpty()) {
                val fallbackElements = doc.select(".post-item, .article, .lesson, .book")
                for (element in fallbackElements) {
                    try {
                        val item = extractContentFromElement(element, category)
                        if (item != null) {
                            items.add(item)
                        }
                    } catch (e: Exception) {
                        // Continue processing
                    }
                }
            }
        } catch (e: Exception) {
            // Return empty list on parsing error
        }
        
        return items
    }
    
    /**
     * Extract content data from individual HTML element
     */
    private fun extractContentFromElement(element: Element, category: String?): ContentItem? {
        // Extract title
        val titleElement = element.select(".entry-title, .post-title, h1, h2, h3, .title").first()
            ?: element.select("a[title]").first()
        val title = titleElement?.text()?.trim() ?: return null
        
        // Extract link
        val linkElement = element.select("a[href]").first()
        val linkUrl = linkElement?.attr("href") ?: ""
        val fullLinkUrl = if (linkUrl.startsWith("http")) linkUrl else "$baseUrl$linkUrl"
        
        // Extract image
        val imgElement = element.select("img[src]").first()
        val imageUrl = imgElement?.attr("src") ?: imgElement?.attr("data-src")
        val fullImageUrl = if (!imageUrl.isNullOrEmpty() && imageUrl.startsWith("http")) {
            imageUrl
        } else if (!imageUrl.isNullOrEmpty()) {
            "$baseUrl$imageUrl"
        } else {
            null
        }
        
        // Extract description
        val descElement = element.select(".post-content, .entry-content, .excerpt, .description, p").first()
        val description = descElement?.text()?.trim()
        
        // Extract author
        val authorElement = element.select(".author, .by-author, .post-author").first()
        val author = authorElement?.text()?.trim()
        
        // Determine content type based on classes and URL patterns
        val type = determineContentType(element, fullLinkUrl, title)
        
        return ContentItem(
            id = generateId(title, fullLinkUrl),
            title = title,
            imageUrl = fullImageUrl,
            linkUrl = fullLinkUrl,
            type = type,
            description = description,
            author = author,
            dateAdded = System.currentTimeMillis(),
            category = category
        )
    }
    
    /**
     * Determine content type based on element attributes and URL
     */
    private fun determineContentType(element: Element, url: String, title: String): ContentType {
        val elementClasses = element.className().lowercase()
        val urlLower = url.lowercase()
        val titleLower = title.lowercase()
        
        return when {
            elementClasses.contains("audio") || urlLower.contains("audio") || titleLower.contains("صوتي") -> ContentType.AUDIO
            elementClasses.contains("book") || urlLower.contains("book") || titleLower.contains("كتاب") -> ContentType.BOOK
            elementClasses.contains("lesson") || urlLower.contains("lesson") || titleLower.contains("درس") -> ContentType.LESSON
            elementClasses.contains("fatwa") || urlLower.contains("fatwa") || titleLower.contains("فتوى") -> ContentType.FATWA
            elementClasses.contains("scholar") || urlLower.contains("scholar") || titleLower.contains("عالم") -> ContentType.SCHOLAR
            else -> ContentType.ARTICLE
        }
    }
    
    /**
     * Generate unique ID for content item
     */
    private fun generateId(title: String, url: String): String {
        return "${title}_${url.hashCode()}"
    }
    
    /**
     * Shutdown executor service
     */
    fun shutdown() {
        executor.shutdown()
    }
}